Boost: Ensure delivery method tester is only tested once per network on Multisites

LiamSarsfield opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #41788

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Ensures cron jobs run only once per network instead of per site.
* Code Organization
  * Moved minify legacy notice logic from wp-js-data-sync.php to Minify class
  * Added unit tests for all new scheduling functionality

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

## Testing Instructions

* Install [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/) plugin for cron job verification
* Have access to both a single-site and multisite WordPress installation
* Ensure Jetpack Boost is installed and the Minify feature was previously active

### 1. Migration Testing
<details>
<summary>Instructions</summary>

#### Verify Legacy Cron Migration
1. **Before Update/Branch/PR**
   - Open WP Crontrol (Tools → Cron Events)
   - Note down any existing cron jobs with these names:
     - `jetpack_boost_minify_cron_cache_cleanup`
     - `jetpack_boost_404_tester_cron`
       - Note: The tester cron may NOT appear for Atomic sites due to #42225. This is ok if site is JN or Atomic.
   - Take a screenshot for comparison if needed

2. **Apply Update/Branch/PR**
   - Install/update to this version of Jetpack Boost
   - It's important that the update method also changes the version number of the plugin

3. **Verify Migration**
   - Check WP Crontrol again
   - Confirm the cron jobs are still present with the same name, but scheduling times may have changed.
</details>

### 2. Multisite Testing
<details>
<summary>Instructions</summary>

#### Network Admin Testing
1. **Network Admin View**
   - Log in as Network Administrator
   - Go to Network Admin Main Site → Jetpack Boost
   - Toggle off all Minify features
   - Check WP Crontrol again
   - Verify that the cron jobs:
     - `jetpack_boost_minify_cron_cache_cleanup` and `jetpack_boost_404_tester_cron` **are no** longer present.
   - Toggle On at least one Minify feature
   - Check WP Crontrol again
   - Verify that the cron jobs are present:
     - `jetpack_boost_minify_cron_cache_cleanup`
     - `jetpack_boost_404_tester_cron
     	- Note: The tester cron may NOT appear for Atomic sites due to #42225. This is ok if site is JN or Atomic.

2. **Individual Site Admin Testing**
   - Log in as a regular site administrator
   - Go to individual site's Jetpack Boost settings
   - Toggle off all Minify features
   - Check WP Crontrol again
   - Verify that the cron jobs:
     - `jetpack_boost_minify_cron_cache_cleanup` and `jetpack_boost_404_tester_cron` **are no** longer present.
   - Toggle On at least one Minify feature
   - Check WP Crontrol again
   - Verify that the cron jobs are present:
     - `jetpack_boost_minify_cron_cache_cleanup`
     - `jetpack_boost_404_tester_cron
     	- Note: The tester cron may NOT appear for Atomic sites due to #42225. This is ok if site is JN or Atomic.

#### Cron Execution Testing
1. **Initial Setup**
   - Create at least 3 subsites in your multisite installation
   - Enable Jetpack Boost on each site
   - Enable a Minify feature on each site

2. **Monitor Cron Execution**
   - In WP Crontrol, observe the `jetpack_boost_minify_cron_cache_cleanup` and`jetpack_boost_404_tester_cron` events.
   - Note the next scheduled run time
   - Use WP Crontrol to manually trigger the cron event
   - Check site option in the database:
     ```sql
     SELECT * FROM `wp_sitemeta` WHERE (`meta_key` LIKE '%last_run%' OR `meta_value` LIKE '%last_run%') LIMIT 50
     ```
 - Verify only one entry cronjob hook exists per cron job, not per site.

3. **Cross-Site Verification**
   - Enable Debug logging
   - Add an `error_log` to the `jetpack_boost_minify_cache_cleanup` function
     - `projects/plugins/boost/app/lib/minify/functions-helpers.php`
   - Add this to wp-config.php:
     ```php
     define('WP_DEBUG', true);
     define('WP_DEBUG_LOG', true);
     ```
   - Toggle Minify JS on Site A
   - Refresh Settings page after a second or so, to ensure the cron has been ran.
   - Verify that an error log has been logged for `jetpack_boost_minify_cache_cleanup`
   - Toggle Minify JS on Site B
   - Verify that an error log has **not** been logged for `jetpack_boost_minify_cache_cleanup`

4. **Recovery Testing**
   - Manually delete cron jobs using WP Crontrol
   - Retoggle all Minify features to off, and then back on
   - Verify cron jobs are properly rescheduled
</details>

### Notes
- Some operations may need to wait for the scheduled time - use WP Crontrol to trigger them manually for immediate testing
- Keep an eye on the debug.log for any unexpected behaviors or errors

LiamSarsfield committed:
Boost: Improve Network-Wide Cron Job Scheduling

- Add `jetpack_boost_execute_network_cron()` to schedule singleton network-wide cron jobs
- Implement `jetpack_boost_schedule_singleton_network_cron()` to ensure unique cron job scheduling
- Update 404 tester cron job scheduling to use new network-wide cron method

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Boost: Improve Multisite and Cron Job Handling

- Add multisite check to prevent legacy notice for non-network admins
- Refactor cron job scheduling for 404 tester and cache cleanup
- Simplify network-wide cron job scheduling logic

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into boost/update/404-tester-multisite-check

LiamSarsfield committed:
Boost: Move Minify Legacy Notice to Minify Module

- Refactor legacy notice logic from wp-js-data-sync.php to Minify.php
- Implement Has_Data_Sync interface for Minify module
- Register readonly data sync entry for minify legacy notice
- Add static method to determine legacy notice visibility

LiamSarsfield committed:
Boost: Improve Network Cron Job Scheduling and Unscheduling

- Modify `jetpack_boost_execute_network_cron()` to dynamically retrieve cron recurrence
- Update `jetpack_boost_schedule_singleton_network_cron()` to save recurrence as a site option
- Add `jetpack_boost_unschedule_singleton_network_cron()` to clean up network cron jobs
- Adjust action hook parameters for network cron scheduling
- Add TODO comment for future code organization

LiamSarsfield committed:
Boost: Refactor Network Cron Job Scheduling and Cleanup

- Move network cron job scheduling and unscheduling logic to a new Scheduled_Event class
- Remove legacy cron job scheduling functions from minify helpers
- Update Jetpack Boost main class to add Scheduled_Event to setup
- Unschedule legacy cronjob names during plugin activation
- Remove unnecessary action hooks and inline cron job functions

LiamSarsfield committed:
Boost: Update Cron Job Scheduling and 404 Tester References

- Update Scheduled_Event class documentation to reference the correct method name
- Replace legacy function calls with Scheduled_Event::schedule_singleton_network_cron()
- Add missing action hook for 404 tester cron job

LiamSarsfield committed:
Boost: Fix Network Cron Job Scheduling Hook

- Update `schedule_singleton_network_cron()` to use a consistent network cron hook
- Modify hook parameters to include original hook and arguments
- Ensure proper tracking of network-wide cron job scheduling

LiamSarsfield committed:
Boost: Enhance Scheduled Event Handling and 404 Tester Cron Job

- Update `schedule_singleton_network_cron()` to return a boolean indicating scheduling status
- Modify 404 tester cron job scheduling to use the updated method
- Simplify cron job scheduling logic in page optimize helpers

LiamSarsfield committed:
Fix incorrect function name

LiamSarsfield committed:
Boost: Add Unit Tests for Scheduled Event Handling

- Create comprehensive test suite for Scheduled_Event class
- Test network cron job scheduling, execution, and unscheduling methods
- Verify behavior for different cron job scenarios using Brain Monkey

LiamSarsfield committed:
Boost: Update Scheduled Event Test Suite to Use Expectation-Based Assertions

- Modify test methods to use `expectNotToPerformAssertions()`
- Remove unnecessary return value checks in test cases
- Simplify test assertions for network cron job methods

LiamSarsfield committed:
Boost: Fix WordPress Action Handling for Empty Arguments

- Add workaround to handle WordPress converting empty arrays to strings for actions
- Ensure actions with no arguments are called correctly using `do_action()`
- Preserve original action call behavior for non-empty argument lists

LiamSarsfield committed:
Boost: Enhance Network Cron Job Tracking and Cleanup

- Add methods to track and manage blogs subscribed to network cron jobs
- Implement `subscribe_to_network_cron()` to record blog participation
- Create `unsubscribe_from_network_cron()` to manage blog subscriptions
- Modify unscheduling logic to clean up network cron options only when no blogs are subscribed

LiamSarsfield committed:
Boost: Update Scheduled Event Test Suite with Blog Subscription Tracking

- Add test expectations for tracking blog subscriptions in network cron jobs
- Implement checks for getting, updating, and deleting blog subscription site options
- Enhance test coverage for network cron job subscription management

LiamSarsfield committed:
Boost: Fix Minify Cache Cleanup with Explicit Type Casting

- Add type checking for `$file_age` parameter in cache cleanup functions
- Ensure consistent default value when non-integer values are passed
- Prevent potential type-related issues during cache cleanup

LiamSarsfield committed:
Merge branch 'trunk' into boost/update/404-tester-multisite-check

haqadn commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -21,6 +22,9 @@ function jetpack_boost_minify_cache_buster() {
  */
 function jetpack_boost_legacy_minify_cache_cleanup( $file_age = DAY_IN_SECONDS ) {
+	// Explicitly cast to int as do_action() can pass a non-int value. https://core.trac.wordpress.org/ticket/14881
+	$file_age = is_int( $file_age ) ? $file_age : DAY_IN_SECONDS;
```
I am a bit confused about this. Don't we want to do `(int) $file_age` instead?
```suggestion
	$file_age = (int) $file_age;
```

LiamSarsfield replied:
My comment was misleading actually, so I've updated it.
`$file_age` becomes an empty string when doing an action that calls `jetpack_boost_legacy_minify_cache_cleanup`, which is the case with the network cron. So essentially:
```
add_action( 'jetpack_boost_minify_cron_cache_cleanup', 'jetpack_boost_legacy_minify_cache_cleanup' );
do_action( 'jetpack_boost_minify_cron_cache_cleanup' );
```
Due to https://core.trac.wordpress.org/ticket/14881

haqadn commented on projects/plugins/boost/app/lib/Scheduled_Event.php:
```diff
@@ -0,0 +1,107 @@
+
+use Automattic\Jetpack_Boost\Contracts\Has_Setup;
+
+class Scheduled_Event implements Has_Setup {
```
`Singleton_Network_Event` would be a clearer name as this class specifically deals with network-level cron events.

LiamSarsfield replied:
Good idea, it compliments your other suggestion regarding the function names.

haqadn commented on projects/plugins/boost/app/lib/Scheduled_Event.php:
```diff
@@ -0,0 +1,107 @@
+	 *
+	 * @return bool True if the cronjob was scheduled, false if it was already scheduled.
+	 */
+	public static function schedule_singleton_network_cron( int $timestamp, string $recurrence, string $hook, array $args = array() ) {
```
I would also advocate for simplifying the names and increase bravity. For example, `Singleton_Network_Event::schedule` seems to convey the same amount of information.

LiamSarsfield replied:
All functions names should be addressed now.

haqadn commented on projects/plugins/boost/app/lib/Scheduled_Event.php:
```diff
@@ -0,0 +1,107 @@
+
+use Automattic\Jetpack_Boost\Contracts\Has_Setup;
+
+class Scheduled_Event implements Has_Setup {
```
This class would also benefit from a detailed PHPdoc

LiamSarsfield replied:
I've added further documentation to the class. Let me know if there's anything still unclear.

donnchawp reviewed (changes requested):
When logged in as a super-admin user I saw that the old cron events were converted to the new style when I upgraded to this patch.
However, when I reloaded the Boost settings page, the old style cron events were added back in, so there were old and new events.

Disabling both modules removed the new style events only. The old style ones remains but the cron_cache_cleanup one didn't have any action, meaning the function wasn't hooked on the action.

I haven't looked at the code yet.

donnchawp commented on projects/plugins/boost/app/lib/Scheduled_Event.php:
```diff
@@ -0,0 +1,107 @@
+
+		update_site_option( "{$action}_network_cron_ran", $current_time );
+
+		do_action( $action, ...$args );
```
Unfortunately, the spread operator is [PHP 7.4](https://wiki.php.net/rfc/spread_operator_for_array) and the plugin only requires PHP 7.2, so this will have to be changed.

LiamSarsfield replied:
@donnchawp My understanding is PHP 7.4 introduced the spread operator for array merging specifically, but for argument unpacking within functions it is supported from PHP 5.6
https://www.php.net/manual/en/migration56.new-features.php
Let me know if I'm misinterpreting that though.

donnchawp replied:
Oh! TIL. It's something I've never used or noticed in PHP code, but now I see it in `do_action`! Thanks!

donnchawp commented on projects/plugins/boost/app/lib/Scheduled_Event.php:
```diff
@@ -0,0 +1,107 @@
+	private static function subscribe_to_network_cron( string $hook ) {
+		$blogs_subscribed                          = get_site_option( "{$hook}_network_cron_blogs_subscribed", array() );
+		$blogs_subscribed[ get_current_blog_id() ] = true;
+		update_site_option( "{$hook}_network_cron_blogs_subscribed", $blogs_subscribed );
```
Unfortunately, this won't scale on very large networks as it'll be passing around really large arrays of blog_ids.

What might work instead:
1. Forget about subscribing to an event.
2. Schedule the job locally on the blog using the jetpack_boost_network_cron event.
3. When the job is unscheduled, it doesn't need to check subscription status. It just unschedules the local job.
4. A plugin deactivate or uninstall function removes any site options left over.
5. The local blog doesn't need to care about other blogs.

LiamSarsfield replied:
@donnchawp I've updated the implementation per your suggestion:

It no longer subscribe the blog to the event. Instead, it simply unsubscribes to the existing blog's event.
When the plugin is uninstalled the site options are removed per below:
https://github.com/Automattic/jetpack/blob/33fcab15ed051740efebe2c028daa46a934a868e/projects/plugins/boost/app/class-jetpack-boost.php#L298-L331

donnchawp reviewed (changes requested):
Some more changes.

haqadn reviewed (changes requested):
I have a few nitpicks about the code. To me, it seems a bit more powerful than it needs to be. Also what @donnchawp mentioned about passing around big list of blog ids make sense.

Can we instead go for an approach where we find a way to designate a single site to handle the network cron? This implementation seems to do pretty much the same thing. 
I am guessing the array of blog IDs is there to make sure we correctly register/unregister the cron when all sites have the minify modules turned off for example.

So my suggestion would be an approach like this:
- Each site (and maybe even each module) can register/unregister cron events independently.
- When running the cron, check when was the last time it ran on the whole network.
- Exit early if it already ran in a given amount of time.

In other words, I am suggesting that instead of having a top level cron that triggers sub-events on each site, we handle the deduplication on the cron itself. Now, this means the same cron will execute a lot of times. But, as long as we don't run the heavy action associated with each cron, I think we should be fine. Unless I am missing something here.

The reason I am hesitant about the current approach is because I find it hard to understand how it works. After going through it a few times, I understand it. But, simpler would be better.

LiamSarsfield committed:
Boost: Improve Minify Cache Cleanup Comment and Type Handling

- Enhance comment clarity for file age parameter type checking

LiamSarsfield committed:
Boost: Refactor Singleton Network Event Handling

- Rename `Scheduled_Event` to `Singleton_Network_Event`
- Simplify network cron event management by removing blog subscription tracking
- Add method to unschedule all network cron events
- Update references across the plugin to use the new class name
- Streamline event scheduling and unscheduling logic

LiamSarsfield committed:
Boost: Improve Version Change Handling with Scheduled Cron Event

- Modify `schedule_version_change()` to use a scheduled cron event for version updates
- Defer version update logic
- Simplify version change detection and update process

LiamSarsfield committed:
Merge branch 'trunk' into boost/update/404-tester-multisite-check

LiamSarsfield committed:
Boost: Add Unit Tests for Singleton Network Event Handling

- Update Test_Scheduled_Event to adhere to new Base_TestCase
- Add detailed PHPDocs for Singleton_Network_Event

LiamSarsfield committed:
Boost: Optimize Version Change Handling Timing

- Reduce cron event delay from 10 to 2 seconds
- Move version update option setting before scheduling the cron event

LiamSarsfield committed:
Boost: Prevent Duplicate Version Change Cron Event

- Add check to prevent scheduling multiple version change cron events
- Move action hook registration outside of schedule_version_change() method
- Ensure only one version change event is scheduled at a time

LiamSarsfield commented:
> Can we instead go for an approach where we find a way to designate a single site to handle the network cron? This implementation seems to do pretty much the same thing.

Unfortunately this may not be reliable as each site maintains independent cron schedules.
If we designate the main/root site as the cron handler for example, the cron job may not run if a multisite is designed in a way where the main/root site is rarely called due to the independent cron schedules.

> I am guessing the array of blog IDs is there to make sure we correctly register/unregister the cron when all sites have the minify modules turned off for example.

Yep, that was the original intention. But due to scalability concerns as mentioned by @donnchawp this may not be optimal. Instead of removing those site options upon unscheduling of the cron, they will be removed if the user uninstalls the plugin instead.

https://github.com/Automattic/jetpack/blob/80295c1e452d1a9f345a7df20f029172c8e04358/projects/plugins/boost/app/class-jetpack-boost.php#L314-L329

> So my suggestion would be an approach like this:
> - Each site (and maybe even each module) can register/unregister cron events independently.
> - When running the cron, check when was the last time it ran on the whole network.
> - Exit early if it already ran in a given amount of time.

Yep, that's essentially what the execute is currently doing now.
Each individual site calls this `execute` that has this multisite cronjob.

https://github.com/Automattic/jetpack/blob/80295c1e452d1a9f345a7df20f029172c8e04358/projects/plugins/boost/app/lib/Singleton_Network_Event.php#L77-L95

LiamSarsfield commented:
@donnchawp 

> When logged in as a super-admin user I saw that the old cron events were converted to the new style when I upgraded to this patch.
> However, when I reloaded the Boost settings page, the old style cron events were added back in, so there were old and new events.

I was able to reproduce this on a Jurassic Ninja site, but the only theory I have on why this was the case was if the previous `handle_version` was being called, even after the scheduling of the new cron and unscheduling of the old cron was done.

I have updated it so the handling of the version change is done in a single event cron that executes 2 seconds after the change is detected. After implementing like this, I'm no longer able to reproduce.
https://github.com/Automattic/jetpack/blob/80295c1e452d1a9f345a7df20f029172c8e04358/projects/plugins/boost/app/class-jetpack-boost.php#L157-L180

donnchawp commented on projects/plugins/boost/app/class-jetpack-boost.php:
```diff
@@ -297,12 +311,26 @@ public function uninstall() {
 				WHERE  `option_name` LIKE 'jetpack_boost_%';
 			"
 		);
+		$site_options = $wpdb->get_col(
```
Let's delete the site options one by one. Direct SQL is discouraged in WordPress plugins, and deleting site options this way feels unsafe.

I see that the plugin already handles local blog options in this way, but we really should delete those options one by one. I'll add an issue about this we can use to track that.

LiamSarsfield replied:
I've updated the PR to address this now. 
All site options are now within 2 options, each with their own key (hook), value (executed timestamp, or recurrence value), so we can delete these site options easier.

donnchawp reviewed (changes requested):
Looks great, and the cron events issue is gone, but my only niggle is with how the site options are deleted. Even though the plugin deletes blog options with an SQL query to fetch them, I think they should be deleted one by one.

LiamSarsfield committed:
Boost: Refactor Singleton Network Event Cron Management

- Introduce constants for network cron executed and recurrence options
- Add methods to get, set, and validate cron execution timestamps and recurrence
- Improve cron event tracking with more robust site option management
- Simplify cron execution and unscheduling logic
- Remove legacy option naming approach in favor of centralized option handling

LiamSarsfield committed:
Boost: Rename and Update Singleton Network Event Test File

- Rename Test_Scheduled_Event.php to Test_Singleton_Network_Event.php
- Update test cases to match recent Singleton_Network_Event refactoring
- Add new test methods for get_cron_executed, get_cron_recurrence, and unschedule_all
- Improve test coverage for network cron event management

LiamSarsfield committed:
Merge branch 'trunk' into boost/update/404-tester-multisite-check

LiamSarsfield committed:
Boost: Fix Cron Schedule Interval Retrieval

Remove unnecessary null coalescing for interval to ensure consistent return of schedule interval

donnchawp reviewed (dismissed):
Works as expected!

LiamSarsfield committed:
Boost: Refactor Network Cron Event Management

Simplify and improve network-wide cron event synchronization:
- Replace complex cron execution tracking with a more straightforward approach
- Modify Singleton_Network_Event to filter and synchronize cron jobs across network
- Update method names and remove unnecessary complexity in cron event handling

LiamSarsfield committed:
Boost: Improve Network Cron Job Synchronization Logic

- Enhance filter_cron_jobs method to handle network-wide cron job synchronization
- Add logic to prevent running cron jobs too early or within a day of Unix time
- Improve cron job rescheduling by unscheduling and rescheduling events
- Update method signatures and variable names for clarity
- Add test coverage for new cron job synchronization behavior

LiamSarsfield committed:
Merge branch 'trunk' into boost/update/404-tester-multisite-check

LiamSarsfield committed:
Satisfy Jetpack.PHPUnit.TestClassName.DoesNotEndWithTest

LiamSarsfield committed:
Remove unused network cron job methods

- Remove `add_action` for 'jetpack_boost_network_cron' in setup method
- Remove static `unschedule` method
- Update corresponding test cases to remove related test methods

LiamSarsfield committed:
Boost: Refine Network Cron Job Synchronization

- Remove legacy cron job unscheduling in Jetpack Boost upgrade
- Remove unnecessary timestamp checks in cron job synchronization
- Update method comments for clarity

LiamSarsfield committed:
Merge branch 'trunk' into boost/update/404-tester-multisite-check

LiamSarsfield committed:
Simplify filter_cron_jobs loop logic

LiamSarsfield committed:
Refactor Singleton_Network_Event to improve cron job handling

- Change `$is_filtering` from static to instance property
- Add action hook for scheduling network cron events
- Simplify cron job filtering and rescheduling logic
- Improve handling of network-wide cron job timestamps

LiamSarsfield committed:
Fix wp_next_scheduled() parameter in Singleton_Network_Event

- Correct the parameter passed to wp_next_scheduled() from $recurrence to $hook
- Update test cases to reflect the correct parameter usage

LiamSarsfield committed:
Improve Singleton_Network_Event cron job scheduling logic

- Refactor schedule() method to handle cron job timestamps more robustly
- Modify logic to check and set cron job execution timestamps
- Update test case to expect multiple get_site_option() calls

LiamSarsfield committed:
Remove Singleton_Network_Event and refactor cron job scheduling

- Delete Singleton_Network_Event class and its associated test file
- Modify minify and 404 tester functions to use standard WordPress cron scheduling
- Add direct site option management for tracking last run of cron jobs
- Simplify network cron job scheduling logic in Jetpack Boost plugin

LiamSarsfield committed:
Remove redundant comment in 404 tester function

- Delete unnecessary comment about updating 404_tester_last_run
- Maintain existing 404 tester function logic

haqadn commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -15,12 +15,31 @@ function jetpack_boost_minify_cache_buster() {
+ * This ensures that the cache cleanup cron job is only run once per day, espicially for multisite.
+ */
+function jetpack_boost_minify_cron_cache_cleanup() {
+	// If we see it's been executed within 24 hours, don't run
+	if ( get_site_option( 'jetpack_boost_minify_cron_cache_cleanup_last_run', 0 ) > time() - DAY_IN_SECONDS ) {
+		return;
+	}
+
+	update_site_option( 'jetpack_boost_minify_cron_cache_cleanup_last_run', time() );
```
Can we extract this into a function?

LiamSarsfield committed:
Refactor daily network cron job scheduling

- Introduce jetpack_boost_should_run_daily_network_cron_job() helper function
- Simplify 404 tester and minify cache cleanup cron job logic
- Remove redundant site option update in individual cron job functions
- Centralize daily network cron job execution check

haqadn reviewed (approved):
Everything seems to work as expected and the changes look good

I couldn't test the 404 tester cron as I was using Jurassic Ninja.
